### PR TITLE
Require explicit redirects and drop www_redirect

### PR DESCRIPTION
**Pull request by fullyint**
Discontinues the `reverse_www` filter, overcoming its challenges in determining when to add/remove www in presence of subdomains.

Some history in #570 and #619.

Example of problem to solve: https://discourse.roots.io/t/https-www-unsecure-and-not-redirecting/7071/

### CODE REVIEW

File: roles/common/templates/site_hosts.j2 [hunk 1]
```diff
@@ -0,0 +1,17 @@
+Required format for `site_hosts` (group_vars/{{ env }}/wordpress_sites.yml):
+
+example.com:
+  site_hosts:
+    - canonical: example.com
+
+The above is the minimum required. Multiple hosts and redirects are posible:
```
Comment (swalkinshaw) [inline roles/common/templates/site_hosts.j2]: `possible`

File: Vagrantfile [hunk 2]
```diff
@@ -49,14 +49,20 @@ Vagrant.configure('2') do |config|
...
+  hostname, *aliases = wordpress_sites.flat_map { |(_name, site)| site['site_hosts'].map { |host| host['canonical'] } }
   config.vm.hostname = hostname
-  www_aliases = ["www.#{hostname}"] + aliases.map { |host| "www.#{host}" }
+  redirects = wordpress_sites.flat_map { |(_name, site)|  site['site_hosts'].select { |host| host.has_key?('redirects') }.flat_map { |host| host['redirects'] } }
 
   if Vagrant.has_plugin? 'vagrant-hostmanager'
     config.hostmanager.enabled = true
     config.hostmanager.manage_host = true
-    config.hostmanager.aliases = aliases + www_aliases
+    config.hostmanager.aliases = aliases + redirects
```
Comment (swalkinshaw) [inline Vagrantfile]:
```ruby
  site_hosts = wordpress_sites.flat_map { |(_name, site)| site['site_hosts'] }

  site_hosts.each do |host|
    if !host.is_a?(Hash) or !host.has_key?('canonical')
      fail_with_message File.read(File.join(ANSIBLE_PATH, 'roles/common/templates/site_hosts.j2')).sub!('{{ env }}', 'development').gsub!(/com$/, 'dev')
    end
  end

  main_hostname, *hostnames = site_hosts.map { |host| host['canonical'] }
  config.vm.hostname = main_hostname

  redirects = site_hosts.flat_map { |host| host['redirects'] }.compact

  if Vagrant.has_plugin? 'vagrant-hostmanager'
    config.hostmanager.enabled = true
    config.hostmanager.manage_host = true
    config.hostmanager.aliases = hostnames + redirects
  else
    fail_with_message "vagrant-hostmanager missing, please install the plugin with this command:\nvagrant plugin install vagrant-hostmanager"
  end
```

File: Vagrantfile [hunk 3]
```diff
@@ -49,14 +49,20 @@ Vagrant.configure('2') do |config|
   # Required for NFS to work
   config.vm.network :private_network, ip: ip, hostsupdater: 'skip'
 
-  hostname, *aliases = wordpress_sites.flat_map { |(_name, site)| site['site_hosts'] }
+  wordpress_sites.flat_map { |(_name, site)| site['site_hosts'] }.each do |host|
+    if !host.is_a?(Hash) or !host.has_key?('canonical')
+      fail_with_message File.read(File.join(ANSIBLE_PATH, 'roles/common/templates/site_hosts.j2')).sub!('{{ env }}', 'development').gsub!(/com$/, 'dev')
```
Comment (swalkinshaw) [inline Vagrantfile]: Match on `\.com` and replace with `.dev`

File: roles/common/tasks/main.yml [hunk 4]
```diff
@@ -10,6 +10,13 @@
   when: ansible_version is not defined or false in [{% for item in ansible_requirements %}{{ ansible_version.full | version_compare(item.version, item.operator) }},{% endfor %}]
   run_once: true
 
+- name: Validate format of site_hosts
+  fail:
+    msg: "{{ lookup('template', 'site_hosts.j2') }}"
+  with_dict: "{{ wordpress_sites }}"
+  when: 0 < [{% for host in item.value.site_hosts if 'canonical' not in host %}'',{% endfor %}] | count
```
Comment (swalkinshaw) [inline roles/common/tasks/main.yml]: `when: item.value.site_hosts | map(attribute='canonical') | count != item.value.site_hosts | count`

Same thing and any better?

Comment (swalkinshaw) [inline roles/common/tasks/main.yml]: Or `when: site_hosts_canonical | count != item.value.site_hosts | count`

Comment (swalkinshaw) [inline roles/common/tasks/main.yml]: scratch these, won't work :(

Comment (fullyint) [inline roles/common/tasks/main.yml]: @swalkinshaw Will use your superior suggestion of 
`when: item.value.site_hosts | rejectattr('canonical', 'defined') | list | count`
Quite beautiful. Thank you.

File: Vagrantfile (hunk 2 above)
Comment (fullyint) [inline Vagrantfile]: Your refactor of `site_hosts` mapping is sooo much better. Thanks!

File: roles/letsencrypt/tasks/certificates.yml [hunk 5]
```diff
@@ -15,19 +15,19 @@
   tags: [letsencrypt_keys]
 
 - name: Generate CSRs for single domain keys
```
Comment (swalkinshaw) [inline roles/letsencrypt/tasks/certificates.yml]: Let's get rid of this and consolidate the tasks into a single CSR one with the SAN feature. I tested it and it works:

```yaml
- name: Generate CSRs
  shell: "openssl req -new -sha256 -key '{{ letsencrypt_keys_dir }}/{{ item.key }}.key' -subj '/' -reqexts SAN -config <(cat /etc/ssl/openssl.cnf <(printf '[SAN]\nsubjectAltName=DNS:{{ site_hosts | join(',DNS:') }}')) > {{ acme_tiny_data_directory }}/csrs/{{ item.key }}.csr"
  args:
    executable: /bin/bash
    creates: "{{ acme_tiny_data_directory }}/csrs/{{ item.key }}.csr"
  when: site_uses_letsencrypt
  with_dict: "{{ wordpress_sites }}"
  tags: [letsencrypt_keys]
```

File: group_vars/all/main.yml [hunk 6]
```diff
@@ -17,6 +17,8 @@ wordpress_env_defaults:
   wp_siteurl: "${WP_HOME}/wp"
 
 site_env: "{{ wordpress_env_defaults | combine(item.value.env | default({}), vault_wordpress_sites[item.key].env) }}"
+site_hosts: "[{% for item in item.value.site_hosts %}'{{ item.canonical }}',{% for redirect in item.redirects | default([]) %}'{{ redirect }}',{% endfor %}{% endfor %}]"
+site_hosts_canonical: "{{ item.value.site_hosts | map(attribute='canonical') | list }}"
```
Comment (swalkinshaw) [inline group_vars/all/main.yml]: Alternative:

```yaml
site_hosts_canonical: "{{ item.value.site_hosts | map(attribute='canonical') | list }}"
site_hosts_redirects: "{{ item.value.site_hosts | selectattr('redirects', 'defined') | map(attribute='redirects') | map('first') | list }}"
site_hosts: "{{ site_hosts_canonical | union(site_hosts_redirects) }}"
```

These have the advantage of making usage in other roles simpler. No need for looping or `default`.

File: roles/letsencrypt/templates/nginx-challenge-site.conf.j2 [hunk 7]
```diff
@@ -1,5 +1,5 @@
 server {
   listen 80;
-  server_name {{ item.item.value.site_hosts | reverse_www(enabled=item.item.value.www_redirect | default(true)) | join(' ') }};
+  server_name{% for item in item.item.value.site_hosts %} {{ item.canonical }}{% for redirect in item.redirects | default([]) %} {{ redirect }}{% endfor %}{% endfor %};
```
Comment (swalkinshaw) [inline roles/letsencrypt/templates/nginx-challenge-site.conf.j2]: Can this be `site_hosts | join(',')`?

File: roles/wordpress-setup/templates/wordpress-site.conf.j2 [hunk 8]
```diff
@@ -95,7 +95,7 @@ server {
 }
 {% endif %}
 
-{% for host in item.value.site_hosts if item.value.www_redirect | default(true) %}
+{% for host in item.value.site_hosts if host.redirects | default([]) %}
```
Comment (swalkinshaw) [inline roles/wordpress-setup/templates/wordpress-site.conf.j2]: Would this work:

`{% for host in site_hosts_redirects %}`

File: roles/letsencrypt/templates/nginx-challenge-site.conf.j2 (hunk 7 above)
Comment (fullyint) [inline roles/letsencrypt/templates/nginx-challenge-site.conf.j2]: The disappointment here is that it uses `item.item.value.site_hosts`.
We would need just `item.value.site_hosts` in order to use `site_hosts` from `helpers.yml`.

I tried opening the template with `{% set item = item.item %}` but it doesn't work.
If it feels cleaner or more consistent, we could open the template with
`{% set canonical = ... [something] %}`
`{% set redirects = ... [something] %}`
`{% set site_hosts = ... [something with union] %}`
and _then_ do the `site_hosts | join(' ')`

Comment (swalkinshaw) [inline roles/letsencrypt/templates/nginx-challenge-site.conf.j2]: Not worth it, good as is 👌

**Comment by swalkinshaw** [comment]
💯 so much better

**Pull request merged**